File: R/CWFIS_active_fires.R
### By Dan Thompson
### Simple GeoServer call to the WFS active fire layer
library(sf)
library(httr2)
library(geojsonsf)

### See https://cwfis.cfs.nrcan.gc.ca/interactive-map for browser version of same layer

### WFS active fires request for  all of Canada as json
urlActiveFires=paste0("http://cwfis.cfs.nrcan.gc.ca/geoserver/public/ows?service=WFS&version=1.0.0&request=GetFeature&typeName=public:activefires_current&outputFormat=json")

#convert json to sf
fires_current_json = request(urlActiveFires) |> req_perform() |> resp_body_string()
  fires_current_sf = geojson_sf(fires_current_json)
    
    ## custom filters of sf dataframe using dplyr
  fires_current <- fires_current_sf %>% ### preserve as sf
  filter(hectares>1000) %>% ##filter only fires >1,000 hectares
  filter(stage_of_control == "OC")
  
  ### see glossary at https://ciffc.ca/ for definition of stage_of_control levels
  
  
  ## can use st_drop_geometry() to strip locations (EPSG:3978) if you want
  ##lat/long is embedded in the data table in EPSG:4326  otherwise

File: py/CWFIS_active_fires.py
'''20230603 query CIFFC active fire locations

NB need to clip Sentinel-2 footprint for Canada'''
from urllib.request import urlopen
from misc import time_stamp, run
import shutil
import json
import sys

data_json = None  # read json data
try:  
    url = 'https://cwfis.cfs.nrcan.gc.ca/geoserver/public/ows?service=WFS&version=1.0.0&request=GetFeature&typeName=public:activefires_current&outputFormat=json'
    response = urlopen(url)
    data_json = response.read()

    # save to file
    ts = time_stamp()
    jfn = ts + '_CWFIS_fires.json'
    open(jfn, 'wb').write(data_json)
    print('+w', jfn)

    # convert to shapefile
    shutil.copyfile(jfn, 'CWFIS_fires.json')
    run('ogr2ogr -skipfailures -f "ESRI Shapefile" CWFIS.shp CWFIS_fires.json')
except:
    data_json = open('CWFIS_fires.json').read()

# parse json data
data_json = json.loads(data_json)
# print(data_json)

select_agency = None
if len(sys.argv) > 1:
    select_agency = sys.argv[1]

agency = {}
for f in data_json['features']:
    print(f.keys()) # dict_keys(['type', 'id', 'geometry', 'geometry_name', 'properties'])
    properties = f['properties']
    ag = properties['agency']
    if ag not in agency:
        agency[ag] = 0
    agency[ag] += 1
    if ag == select_agency or select_agency is None:
        print(properties)

print(agency)
